server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentAction.java

```diff
@@ -0,0 +1,251 @@
+      if (additionalFields.contains(ADDITIONAL_PERIODS)) {
+        response.getPeriodsBuilder().addAllPeriods(periods);
+      }
+    }
+
+    return response.build();
+  }
+
+  private List<MetricDto> searchMetrics(DbSession dbSession, ComponentWsRequest request) {
+    List<MetricDto> metrics = dbClient.metricDao().selectByKeys(dbSession, request.getMetricKeys());
+    if (metrics.size() < request.getMetricKeys().size()) {
+      List<String> foundMetricKeys = Lists.transform(metrics, MetricDtoFunctions.toKey());
+      Set<String> missingMetricKeys = Sets.difference(
+        new LinkedHashSet<>(request.getMetricKeys()),
+        new LinkedHashSet<>(foundMetricKeys));
+
+      throw new NotFoundException(format("The following metric keys are not found: %s", Joiner.on(", ").join(missingMetricKeys)));
+    }
+
+    return metrics;
+  }
+
+  private List<MeasureDto> searchMeasures(DbSession dbSession, ComponentDto component, @Nullable SnapshotDto snapshot, List<MetricDto> metrics, List<WsMeasures.Period> periods) {
+    if (snapshot == null) {
+      return emptyList();
+    }
+
+    List<Integer> metricIds = Lists.transform(metrics, MetricDtoFunctions.toId());
+    List<MeasureDto> measures = dbClient.measureDao().selectBySnapshotIdsAndMetricIds(dbSession, singletonList(snapshot.getId()), metricIds);
+    addBestValuesToMeasures(measures, component, metrics, periods);
+
+    return measures;
+  }
+
+  /**
+   * Conditions for best value measure:
+   * <ul>
+   * <li>component is a production file or test file</li>
+   * <li>metric is optimized for best value</li>
+   * </ul>
+   */
+  private static void addBestValuesToMeasures(List<MeasureDto> measures, ComponentDto component, List<MetricDto> metrics, List<WsMeasures.Period> periods) {
+    if (!QUALIFIERS_ELIGIBLE_FOR_BEST_VALUE.contains(component.qualifier())) {
+      return;
+    }
+
+    List<MetricDtoWithBestValue> metricWithBestValueList = from(metrics)
+      .filter(MetricDtoFunctions.isOptimizedForBestValue())
+      .transform(new MetricDtoToMetricDtoWithBestValueFunction(periods))
+      .toList();
+    Map<Integer, MeasureDto> measuresByMetricId = Maps.uniqueIndex(measures, MeasureDtoFunctions.toMetricId());
+
+    for (MetricDtoWithBestValue metricWithBestValue : metricWithBestValueList) {
+      if (measuresByMetricId.get(metricWithBestValue.getMetric().getId()) == null) {
+        measures.add(metricWithBestValue.getBestValue());
+      }
+    }
+  }
+
+  private static ComponentWsRequest toComponentWsRequest(Request request) {
+    ComponentWsRequest componentWsRequest = new ComponentWsRequest()
+      .setComponentId(request.param(PARAM_COMPONENT_ID))
+      .setComponentKey(request.param(PARAM_COMPONENT_KEY))
+      .setAdditionalFields(request.paramAsStrings(PARAM_ADDITIONAL_FIELDS))
+      .setMetricKeys(request.mandatoryParamAsStrings(PARAM_METRIC_KEYS));
+    checkRequest(!componentWsRequest.getMetricKeys().isEmpty(), "At least one metric key must be provided");
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) NullPointerException might be thrown as 'componentWsRequest' is nullable here [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2259)